### PR TITLE
[LF] use stable Tuple for fecth_by_key primitive

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**

### CODE REVIEW

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 1]
```diff
@@ -1410,9 +1418,10 @@ class TypingSpec(majorLanguageVersion: LanguageMajorVersion)
       val pkgIface = PackageInterface(Map(defaultPackageId -> pkg))
 
       def checkModule(modName: String): Unit = Typing.checkModule(
-        pkgIface,
-        defaultPackageId,
-        pkg.modules(DottedName.assertFromString(modName)),
+        stablePackages = StablePackagesV2,
```
Comment (paulbrauner-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:1421]:
```suggestion
        stablePackages = StablePackages(majorLanguageVersion),
```

File: sdk/compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Context.scala [hunk 2]
```diff
@@ -130,12 +131,17 @@ class Context(
...
-    val compiler = new Compiler(interface, compilerConfig)
+    val pkgInterface = new language.PackageInterface(this.allSignatures)
+    val compiler = new Compiler(pkgInterface, compilerConfig)
 
     modulesToCompile.foreach { mod =>
       if (!omitValidation)
-        assertRight(Validation.checkModule(interface, homePackageId, mod).left.map(_.pretty))
+        assertRight(
+          Validation
+            .checkModule(StablePackagesV2, pkgInterface, homePackageId, mod)
```
Comment (paulbrauner-da) [inline sdk/compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Context.scala:141]:
```suggestion
            .checkModule(StablePackages(languageVersion.major), pkgInterface, homePackageId, mod)
```

File: sdk/daml-lf/encoder/src/test/scala/com/digitalasset/daml/lf/testing/archive/EncodeSpec.scala [hunk 3]
```diff
@@ -279,10 +287,16 @@ object EncodeSpec {
     normalizer.apply(pkg)
   }
 
-  private def validate(pkgId: PackageId, pkg: Package): Unit =
+  private def validate(pkgId: PackageId, pkg: Package): Unit = {
     Validation
-      .checkPackage(language.PackageInterface(Map(pkgId -> pkg)), pkgId, pkg)
+      .checkPackage(
+        StablePackagesV2,
+        language.PackageInterface(StablePackagesV2.packagesMap + (pkgId -> pkg)),
```
Comment (paulbrauner-da) [inline sdk/daml-lf/encoder/src/test/scala/com/digitalasset/daml/lf/testing/archive/EncodeSpec.scala:294]:
```suggestion
        language.PackageInterface(StablePackages(languageVersion.major).packagesMap + (pkgId -> pkg)),
```

File: sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala [hunk 4]
```diff
@@ -597,7 +597,7 @@ class Engine(val config: EngineConfig) {
           // we trust already loaded packages
           .collect {
             case (pkgId, pkg) if !compiledPackages.contains(pkgId) =>
-              Validation.checkPackage(pkgInterface, pkgId, pkg)
+              Validation.checkPackage(StablePackagesV2, pkgInterface, pkgId, pkg)
```
Comment (paulbrauner-da) [inline sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:600]:
```suggestion
              Validation.checkPackage(StablePackages(pkg.languageVersion.major), pkgInterface, pkgId, pkg)
```

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 5]
```diff
@@ -453,7 +453,7 @@ private[lf] final class Compiler(
     config.packageValidation match {
       case Compiler.NoPackageValidation =>
       case Compiler.FullPackageValidation =>
-        Validation.checkPackage(pkgInterface, pkgId, pkg).left.foreach(throw _)
+        Validation.checkPackage(StablePackagesV2, pkgInterface, pkgId, pkg).left.foreach(throw _)
```
Comment (paulbrauner-da) [inline sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:456]:
```suggestion
        Validation.checkPackage(StablePackages(pkg.languageVersion.major), pkgInterface, pkgId, pkg).left.foreach(throw _)
```

File: sdk/daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 6]
```diff
@@ -50,9 +50,14 @@ class TestTraceLog extends TraceLog {
   def getMessages: Seq[String] = messages.view.map(_._1).toSeq
 }
 
-class EvaluationOrderTest_V2 extends EvaluationOrderTest(LanguageVersion.v2_dev)
+class EvaluationOrderTest_V2 extends EvaluationOrderTest(LanguageVersion.v2_dev) {
+  protected def tuple2TyCon: String = {
```
Comment (paulbrauner-da) [inline sdk/daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:54]: You could probably define this in the EvaluationOrderTest class by calling StablePackages(languageVersion.major).

File: sdk/daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTestLib.scala [hunk 7]
```diff
@@ -205,7 +206,7 @@ private[speedy] object SpeedyTestLib {
         s"these packages don't have the expected major language version $majorLanguageVersion: $wrongPackages"
       },
     )
-    Validation.unsafeCheckPackages(PackageInterface(pkgs), pkgs)
+    Validation.unsafeCheckPackages(StablePackagesV2, PackageInterface(pkgs), pkgs)
```
Comment (paulbrauner-da) [inline sdk/daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTestLib.scala:209]:
```suggestion
    Validation.unsafeCheckPackages(StablePackages(majorLanguageVersion), PackageInterface(pkgs), pkgs)
```

File: sdk/daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTestLib.scala [hunk 8]
```diff
@@ -220,7 +221,7 @@ private[speedy] object SpeedyTestLib {
   ): PureCompiledPackages =
     typeAndCompile(
       pkg.languageVersion.major,
-      Map(parserParameter.defaultPackageId -> pkg),
+      StablePackagesV2.packagesMap + (parserParameter.defaultPackageId -> pkg),
```
Comment (paulbrauner-da) [inline sdk/daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTestLib.scala:224]:
```suggestion
      StablePackages(pkg.languageVersion.major).packagesMap + (parserParameter.defaultPackageId -> pkg),
```

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 9]
```diff
@@ -1660,9 +1669,10 @@ class TypingSpec(majorLanguageVersion: LanguageMajorVersion)
 
       def checkModule(pkg: Package, modName: String) =
         Typing.checkModule(
-          pkgInterface,
-          defaultPackageId,
-          pkg.modules(DottedName.assertFromString(modName)),
+          stablePackages = StablePackagesV2,
```
Comment (paulbrauner-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:1672]:
```suggestion
          stablePackages = StablePackages(majorLanguageVersion),
```

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 10]
```diff
@@ -1760,9 +1770,10 @@ class TypingSpec(majorLanguageVersion: LanguageMajorVersion)
       """
 
       def checkModule(pkg: Package, modName: String) = Typing.checkModule(
-        PackageInterface(Map(defaultPackageId -> pkg)),
-        defaultPackageId,
-        pkg.modules(DottedName.assertFromString(modName)),
+        stablePackages = StablePackagesV2,
```
Comment (paulbrauner-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:1773]:
```suggestion
        stablePackages = StablePackages(majorLanguageVersion),
```

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 11]
```diff
@@ -1791,7 +1802,12 @@ class TypingSpec(majorLanguageVersion: LanguageMajorVersion)
       """
 
       val mod = pkg.modules(DottedName.assertFromString("TypeVarShadowing2"))
-      Typing.checkModule(PackageInterface(Map(defaultPackageId -> pkg)), defaultPackageId, mod)
+      Typing.checkModule(
+        stablePackages = StablePackagesV2,
```
Comment (paulbrauner-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:1806]:
```suggestion
        stablePackages = StablePackages(majorLanguageVersion),
```

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 12]
```diff
@@ -1880,7 +1901,12 @@ class TypingSpec(majorLanguageVersion: LanguageMajorVersion)
 
       def checkModule(mod: Module) = {
         val pkg = Package.build(List(mod), List.empty, defaultLanguageVersion, packageMetadata)
-        Typing.checkModule(PackageInterface(Map(defaultPackageId -> pkg)), defaultPackageId, mod)
+        Typing.checkModule(
+          stablePackages = StablePackagesV2,
```
Comment (paulbrauner-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:1905]:
```suggestion
          stablePackages = StablePackages(majorLanguageVersion),
```

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 13]
```diff
@@ -1903,7 +1929,12 @@ class TypingSpec(majorLanguageVersion: LanguageMajorVersion)
...
```
Comment (paulbrauner-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:1933]:
```suggestion
          stablePackages = StablePackages(majorLanguageVersion),
```

File: sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 14]
```diff
@@ -2002,7 +2033,12 @@ class TypingSpec(majorLanguageVersion: LanguageMajorVersion)
 
        }
      """
-    Typing.Env(LV.default, PackageInterface(Map(defaultPackageId -> pkg)), Context.None)
+    Typing.Env(
+      LV.default,
+      StablePackagesV2,
```
Comment (paulbrauner-da) [inline sdk/daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:2038]:
```suggestion
      StablePackages(majorLanguageVersion),
```

**Review by paulbrauner-da — APPROVED** [review]
Trying to think of the poor person who's going to introduce LF 3 in 10 years ;)

I haven't checked that my suggestion actually compile.

**Comment by remyhaemmerle-da** [comment]
> Trying to think of the poor person who's going to introduce LF 3 in 10 years ;)

**Pull request merged**